PRE-LECTURE HW BELOW

In [49]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()


row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

In [50]:
# Assuming your DataFrame is named df
rows, columns = df.shape

print(f"The DataFrame has {rows} rows and {columns} columns.")


The DataFrame has 391 rows and 11 columns.


Observations are similar to the singular 'objects' of data. In my dataset, an individual villager is an observation which has its own variables assigned to it(also called attributes)

Variables is the data which stores information about each observation(villager). (eg. name, gender, species, etc.)

In [53]:
df.describe()

row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

In [54]:
df['species'].value_counts()

species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear         15
mouse        15
horse        15
bird         13
penguin      13
sheep        13
elephant     11
wolf         11
ostrich      10
deer         10
eagle         9
gorilla       9
chicken       9
koala         9
goat          8
hamster       8
kangaroo      8
monkey        8
anteater      7
hippo         7
tiger         7
alligator     7
lion          7
bull          6
rhino         6
cow           4
octopus       3
Name: count, dtype: int64

In [55]:
df.head

<bound method NDFrame.head of      row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
..     ...       ...       ...     ...        ...      ...         ...   
386    475    winnie    Winnie  female      horse     1-31       peppy   
387    477  wolfgang  Wolfgang    male       wolf    11-25      cranky   
388    480      yuka      Yuka  female      koala     7-20      snooty   
389    481      zell      Zell    male       deer      6-7        smug   
390    483    zucker    Zucker    male    octopus      3-8        lazy   

                song    phrase            full_id  \
0         Steep Hill   aye a

In [56]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

titanic = pd.read_csv(url)

titanic.shape

(891, 15)

In [57]:
titanic.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

The key difference between the .shape function and the .describe() function is likely due to missing values.

the shape function includes rows with missing values, while the .describe() function only includes rows not containing missing values

.shape is an attribute, and thus it simply returns the (num rows, num columns)
Note: shape doesnt peform a computation
.describe() is a method, which performs a computation and returns a 'statistical summary of the DataFrame'
Note: methods perform computations, and thus require execution/invokation
even though both return data, methods use () syntax, and the .describe method simply takes no parameters. 

 Here's a summary of our chats:

 1. **Dataset Overview:**
    - You provided a dataset for Animal Crossing villagers and wanted to understand how many rows and columns it has using `.shape` and `.describe()`.

 2. **Observations and Variables:**
    - We discussed that observations are individual rows in your dataset, representing unique instances (villagers). Variables are columns in the dataset, representing different attributes (e.g., name, species, birthday).

 3. **Using `.describe()` and `value_counts()`:**
    - `.describe()` provides summary statistics for numeric columns and can be modified to include categorical columns by using `include='object'`.
    - `df['column'].value_counts()` helps summarize categorical columns by counting the occurrences of each unique value.

 4. **Discrepancy Between `.shape` and `.describe()`:**
    - The discrepancy arises because `.shape` shows the total number of rows and columns, including those with missing values. In contrast, `.describe()` excludes rows with missing values in numeric columns and provides summaries based on non-null data.

 5. **Attributes vs. Methods:**
- We explained that attributes (like `df.shape`) represent data about the object and do not need parentheses, while methods (like `df.describe()`) perform actions and computations, hence require parentheses.

If you have any more questions or need further clarification, feel free to ask!

LINK TO GPT CONVERSATION FOR PRE-LECTURE ASSIGNMENT 

https://chatgpt.com/share/303353cd-ae61-4dca-b97f-0b1917f5ea11


 POST-LECTURE HW BELOW

Count -- The number of non-missing values within a column, shows the COUNT of entries which exist for this variable
Mean -- The mean value represents the average, calculated by summing all values, divided by the count of non-missing values  
std -- The measure of the deviation from the mean value. Higher std means greater data variation from the mean
min -- The min(imum) value within the column, gives the smallest recorded values
25% -- The quartile represents the value in which 25% of the data exists, Q1 represents the 25th percentile of distribution
50% -- The median, or Q2 represents the midpoint of the data, with 50% of the data above and 50% below this value
75% -- The third quartile or Q3 representing the value where 75% of the data exists below. 75th percentile of data distribution
max -- The largest value recorded within the column


7.1) you might use .dropna for customer responses which do not completely respond to all questions, in which you dont want the incomplete responses

7.2) you might no longer wish to hold a row (ie legistlation forces you to stop storing race in a job application database)

7.3) Using del df['col'] will help to remove any unrelated data before clearing the rows or columns that have missing values. doing del df['col'] first may prevent data from being erroneously removed

7.4) 
Check for missing data in each column

In [68]:
missing_data = df.isnull().sum()
print(missing_data)

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64


row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0

we will remove the song column which has significant amounts of missing data

In [69]:
# Before cleaning
print(df.shape)
# Output:(391, 11)

print(df.isnull().sum())

# Delete the 'song' column
del df['song']

# We will remove all other empty rows after deleting irrelevant column. 
# This ensures all empty data is removed 

# Remove rows with any remaining missing data
df_cleaned = df.dropna()

# After cleaning
print(df_cleaned.shape)
print(df_cleaned.isnull().sum())

(391, 11)
row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64
(390, 10)
row_n          0
id             0
name           0
gender         0
species        0
birthday       0
personality    0
phrase         0
full_id        0
url            0
dtype: int64


(390, 10)
row_n          0
id             0
name           0
gender         0
species        0
birthday       0
personality    0
phrase         0
full_id        0
url            0

Here’s a summary of our chats so far:

1. **Dataset Overview and Missing Data Handling:**
   - You provided a dataset for Animal Crossing villagers and wanted to explore handling missing data. We discussed how to view the size of the DataFrame using `.shape` and how to summarize columns with `.describe()` and `df['column'].value_counts()`.

2. **Observations and Variables:**
   - We clarified that **observations** are individual rows (each representing a villager), and **variables** are the columns that describe attributes of each villager (e.g., name, species, gender).

3. **Discrepancies Between `.shape` and `.describe()`:**
   - We discussed why the total number of rows in `.shape` might differ from the count shown in `.describe()`. Missing values in specific columns can cause `.describe()` to exclude some rows, leading to a discrepancy.

4. **Attributes vs. Methods:**
   - I explained why attributes (like `df.shape`) don’t use parentheses, while methods (like `df.describe()`) do. Attributes store data, whereas methods perform actions or calculations.

5. **Handling Missing Data:**
   - We explored how to handle missing data using `df.dropna()` to remove rows or columns with missing values, and `del df['col']` to delete specific columns.
   - We discussed how to clean a dataset by deleting irrelevant columns with missing data and using `dropna()` to remove rows with missing values.
   - You were encouraged to report a "before and after" of the dataset cleaning process, showing the impact of your changes.

Let me know if you have any further questions or need clarification on any points!

LINK TO GPT CONVERSATION FOR POST-LECTURE ASSIGNMENT 

https://chatgpt.com/share/303353cd-ae61-4dca-b97f-0b1917f5ea11

In [70]:
import pandas as pd

url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
titanic_df = pd.read_csv(url)
print(titanic_df.head())


   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True  


In [88]:
titanic_df.groupby("survived")["Fare"].describe()


KeyError: 'Column not found: Fare'

8.1) When grouping by survived, I am looking at the unique values in the column, and considering the rows with the same values into a single group. 
8.1) After grouping by my survived column, im looking at the "fare" column from my grouped data 
Here I am grouping survivors, and non-survivors and then grouping them based on fare costs and describing the dataset.

8.2) When i call .describe, I am getting a summary for each column within my DataFrame, count represents the number of existing values within each column, where if a value is missing in my column, th ecount is less than the total number of rows 
When i use groupby("col1")["col2"].describe(), im getting the count of the existing values within each group defined in "col1". 
this is different because when I calculate my count, it is separately calculated for each group 

8.3) 
a) ChatGPT provides a faster solution to the error -- although very close, ChatGPT immediately gave the code, whereas a stack overflow thread described an unrelated issue 
b) ChatGPT provides a faster solution as it gave multiple ways to troubleshoot the issue, and the code which was relevant. 
c) ChatGPT provided a faster solution -- naive troubleshooting of copypasting resulted in google not giving a relevant result
d) ChatGPT provided a faster solution, even fixing the incorrect code for me
e) ChatGPT troubleshooted faster, fixing the code and explaining my error
f) ChatGPT highlighted a case sensitivity error immediately and fixed it faster
g) ChatGpt was able to quickly solve this issue writing the correct syntax


9) somewhat 

Sure! Here’s a summary of our discussions:

1. **Titanic Dataset Analysis**:
   - You’ve been working with the Titanic dataset and trying to perform various analyses.
   - Issues encountered included:
     - Using incorrect column names or case sensitivity (e.g., `'Survived'` vs. `'survived'`, `'Fare'` vs. `'fare'`).
     - Syntax errors and missing imports, which were resolved by ensuring correct column names and using proper syntax.

2. **Handling Errors**:
   - **NameError**: Occurred when trying to use a variable or column name that was not defined or not recognized. Solution involved checking and using the exact column names.
   - **FileNotFoundError**: Addressed by providing the correct file path or using a URL to load data.
   - **KeyError**: Fixed by verifying and using the exact column names present in the DataFrame.
   - **AttributeError**: Fixed by using the correct method (`.describe()` instead of `.describle()`).

3. **Code Corrections**:
   - Ensured correct syntax for pandas operations like `groupby`, selecting columns, and using `.describe()`.
   - Addressed issues related to case sensitivity and proper use of string literals for column names.

If you have any more questions or need further clarification, feel free to ask!

https://chatgpt.com/share/eb8febf0-92eb-4c32-8adb-e0d29d46b88d